pkg/util/pki/parse.go

```diff
@@ -140,3 +140,179 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
+				continue
+			}
+			if certs[i].Equal(certs[j]) {
+				certs = append(certs[:i], certs[i+1:]...)
+			}
+		}
+	}
+
+	// A certificate chain can be well described as a linked list. Here we build
+	// multiple lists that contain a single node, each being a single certificate
+	// that was passed.
+	var chains []*chainNode
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
+
+	// The task is to build a single list which represents a single certificate
+	// chain. The strategy is to iteratively attempt to join items in the list to
+	// build this single chain. Once we have a single list, we have built the
+	// chain. If the number of lists do not decrease after a pass, then the list
+	// can never be reduced to a single chain and we error.
+	for {
+		// If a single list is left, then we have built the entire chain. Stop
+		// iterating.
+		if len(chains) == 1 {
+			break
+		}
+
+		// lastChainsLength is used to ensure that at every pass, the number of
+		// tested chains gets smaller.
+		lastChainsLength := len(chains)
+		for i := 0; i < len(chains)-1; i++ {
+			for j := 1; j < len(chains); j++ {
+				if i == j {
+					continue
+				}
+
+				// attempt to add both chain together
+				chain, ok := chains[i].tryMergeChain(chains[j])
+				if ok {
+					// If adding the chains together was successful, remove inner chain from
+					// list
+					chains = append(chains[:j], chains[j+1:]...)
+				}
+
+				chains[i] = chain
+			}
+		}
+
+		// If no chains were merged in this pass, the chain can never be built as a
+		// single list. Error.
+		if lastChainsLength == len(chains) {
+			return PEMBundle{}, errors.NewInvalidData("certificate chain is malformed or broken")
+		}
+	}
+
+	// There is only a single chain left at index 0. Return chain as PEM.
+	return chains[0].toBundleAndCA()
+}
+
+// toBundleAndCA will return the PEM bundle of this chain.
+func (c *chainNode) toBundleAndCA() (PEMBundle, error) {
+	var (
+		certs []*x509.Certificate
+		ca    *x509.Certificate
+	)
+
+	for {
+		// If the issuer is nil, we have hit the root of the chain. Assign the CA
+		// to this certificate and stop traversing.
+		if c.issuer == nil {
+			ca = c.cert
+			break
+		}
+
+		// Add this node's certificate to the list at the end. Ready to check
+		// next node up.
+		certs = append(certs, c.cert)
+		c = c.issuer
+	}
+
+	caPEM, err := EncodeX509(ca)
+	if err != nil {
+		return PEMBundle{}, err
+	}
+
+	// If no certificates parsed, then CA is the only certificate and should be
+	// the chain
+	if len(certs) == 0 {
+		return PEMBundle{ChainPEM: caPEM}, nil
+	}
+
+	// Encode full certificate chain
+	chainPEM, err := EncodeX509Chain(certs)
+	if err != nil {
+		return PEMBundle{}, err
+	}
+
+	// Return chain and ca
+	return PEMBundle{CAPEM: caPEM, ChainPEM: chainPEM}, nil
+}
+
+// tryMergeChain will attempt to add the chain "c" and "chain" together.
+// Successful if the either chain is added to the top or bottom of the other
+// chain.
```

Comment:
suggestion: I really liked the way you described what "tryMergeChain" does in https://github.com/jetstack/cert-manager/pull/3978#discussion_r629270243. I propose to add it to the function itself.

```suggestion
// tryMergeChain glues two chains X and Y together by adding one on
// top of the other. The function tries both gluing X on top of Y and
// Y on top of X, which is why the argument order for the two input
// chains does not matter.
//
// As long as both input chains meet conditions K1 and K2, these
// two conditions also hold for the returned chain. K1 and K2 are:
//
// (K1) For all certificates C (child) in the chain, there exists at
// least one certificate P (parent) picked either in the chain itself
// or in the other chain that satisfies C.CheckSignatureFrom(P). In
// the following diagram, C.CheckSignatureFrom(P) is satisfied, i.e.,
// the signature ("sig") on the certificate C can be verified using
// the parent P.
//
// +------+-------+      +------+-------+      +------+-------+
// |      |       |      |      |       |      |      |       |
// |      |  sig ------->|  C   |  sig ------->|  P   |       |
// |      |       |      |      |       |      |      |       |
// +------+-------+      +------+-------+      +------+-------+
// leaf certificate                            root certificate
//
// (K2) The chain is connected, i.e., it is possible to traverse the
// whole chain using the relation "C.CheckSignatureFrom(P) holds".
//
// The function returns false if the above condition does not hold for
// one of the input chains.
```

It would be great to prove that K1 and K2 are sufficient and necessary in both chain so that the resulting chain also satisfies K1 and K2... but I can't bother proving that lol